### PR TITLE
[ZK filter] Remove EnvoyException on data plane

### PR DESCRIPTION
**Pull request by Winbobob**
Commit Message: [ZK filter] Remove EnvoyException on data plane
Additional Description: The Envoy style guideline suggests "Exceptions are allowed on the control plane, though now discouraged in new code. Adding exceptions is disallowed on the data plane", details in https://github.com/envoyproxy/envoy/issues/10878. This diff tries to remove EnvoyException on the ZK proxy filter data plane and paves path for per-opcode decoding error metrics.
Risk Level: Low
Testing: Unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by Winbobob** [comment]
Need to add unit tests to increase the test coverage.

**Comment by Winbobob** [comment]
Hi @JuniorHsu, please check whenever you have a chance, thanks!

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 1]
```diff
@@ -601,87 +909,116 @@ void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, DecodeType dtype
...
+        goto exit_decode_loop;
+      }
+    case DecodeType::WRITE:
+      opcode = decodeOnWrite(data, offset);
+      if (opcode.ok()) {
+        callbacks_.onResponseBytes(opcode.value(), offset - current);
         break;
+      } else {
+        ENVOY_LOG(debug, "zookeeper_proxy: decodeOnWrite exception: {}", opcode.status().message());
+        goto exit_decode_loop;
```
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.cc:948]: let's prevent `goto` by `return`

**Review by JuniorHsu — CHANGES_REQUESTED** [review]
i need more time to deliberate on the macro but put an immediate comment first

File: source/extensions/filters/network/zookeeper_proxy/utils.h [hunk 2]
```diff
@@ -27,20 +29,53 @@ class BufferHelper : public Logger::Loggable<Logger::Id::filter> {
...
+  if (!status.ok()) {                                                                              \
+    return status;                                                                                 \
+  }
+
+#define WRITE_LOG_IF_STATUS_NOT_OK(status, log_level, ...)                                         \
+  if (!status.ok()) {                                                                              \
+    ENVOY_LOG(log_level, ##__VA_ARGS__);                                                           \
+  }
+
+#define COUNT_DECODER_ERR_AND_RETURN_IF_STATUS_NOT_OK(status)                                      \
```
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/utils.h:56]: what does the prefix `COUNT` mean?

File: source/extensions/filters/network/zookeeper_proxy/utils.h [hunk 3]
```diff
@@ -27,20 +29,53 @@ class BufferHelper : public Logger::Loggable<Logger::Id::filter> {
...
+#define COUNT_DECODER_ERR_WITH_LOG_AND_RETURN_IF_STATUS_NOT_OK(status, log_level, ...)             \
+  if (!status.ok()) {                                                                              \
+    ENVOY_LOG(log_level, ##__VA_ARGS__);                                                           \
+    callbacks_.onDecodeError();                                                                    \
+    return status;                                                                                 \
+  }
+
+#define RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status, message)                                   \
+  if (!status.ok()) {                                                                              \
+    return absl::InvalidArgumentError(message);                                                    \
```
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/utils.h:71]: could we just return `status` then we can save a pair of macros?

**Review by JuniorHsu — CHANGES_REQUESTED** [review]
let's try to reduce the number of macro

if macros are zk specific, probably introduce a enum
```
enum class ReportPolicy {
 // comment 
 None, 
 // comment
 DecodeError
}

```
then `ReportPolicy` could be a parameter of macro
let's make ENVOY_LOG always happen in debug level if it applies

File: source/extensions/filters/network/zookeeper_proxy/utils.h (hunk 2 above)
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/utils.h:56]: `callbacks_.onDecodeError();` will increase the decode error counter.

File: source/extensions/filters/network/zookeeper_proxy/utils.h (hunk 3 above)
Comment (Winbobob) [reply source/extensions/filters/network/zookeeper_proxy/utils.h:59]: If the function return type is `absl::Status`, we could return `status` directly.
If the function return type is `absl::StatusOr<T>`, we cannot return `status` directly, so I make it return `absl::InvalidArgumentError` instead.

**Comment by Winbobob** [comment]
> let's try to reduce the number of macro
> 
> if macros are zk specific, probably introduce a enum
> 
> ```
> enum class ReportPolicy {
>  // comment 
>  None, 
>  // comment
>  DecodeError
> }
> ```
> 
> then `ReportPolicy` could be a parameter of macro let's make ENVOY_LOG always happen in debug level if it applies

I tried, but had some issues.

After adding the above enum, I converted the marcos sth like below
```
#define ABSL_STATUS_RETURN_IF_STATUS_NOT_OK(status, report_policy)                                 \
  if (!status.ok()) {                                                                              \
    if (report_policy == ReportPolicy::DecodeError) {                                              \
      callbacks_.onDecodeError();                                                                  \
    }                                                                                              \
    return status;                                                                                 \
  }

#define RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status, report_policy, message)                    \
  if (!status.ok()) {                                                                              \
    if (report_policy == ReportPolicy::DecodeError) {                                              \
      callbacks_.onDecodeError();                                                                  \
    }                                                                                              \
    return absl::InvalidArgumentError(message);                                                    \
  }
```

`utils.cc` also needs to access these marcos, but it cannot identify `callbacks_`, which defines in `decode.h`.
Since `decode.h` already imported `utils.h`, if I try to make `utils.h` import `decode.h`, there will be circular dependency.

File: source/extensions/filters/network/zookeeper_proxy/decoder.cc [hunk 4]
```diff
@@ -601,87 +903,115 @@ void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, DecodeType dtype
...
+    switch (dtype) {
+    case DecodeType::READ:
+      opcode = decodeOnData(data, offset);
+      if (opcode.ok()) {
+        callbacks_.onRequestBytes(opcode.value(), offset - current);
         break;
-      case DecodeType::WRITE:
-        opcode = decodeOnWrite(data, offset);
-        callbacks_.onResponseBytes(opcode, offset - current);
+      } else {
```
Comment (JuniorHsu) [inline source/extensions/filters/network/zookeeper_proxy/decoder.cc:931]: remove the else clause here and below since return after else is an anti-pattern

**Comment by JuniorHsu** [comment]
> > let's try to reduce the number of macro
> > if macros are zk specific, probably introduce a enum
> > ```
> > enum class ReportPolicy {
> >  // comment 
> >  None, 
> >  // comment
> >  DecodeError
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then `ReportPolicy` could be a parameter of macro let's make ENVOY_LOG always happen in debug level if it applies
> 
> I tried, but had some issues.
> 
> After adding the above enum, I converted the marcos sth like below
> 
> ```
> #define ABSL_STATUS_RETURN_IF_STATUS_NOT_OK(status, report_policy)                                 \
>   if (!status.ok()) {                                                                              \
>     if (report_policy == ReportPolicy::DecodeError) {                                              \
>       callbacks_.onDecodeError();                                                                  \
>     }                                                                                              \
>     return status;                                                                                 \
>   }
> 
> #define RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status, report_policy, message)                    \
>   if (!status.ok()) {                                                                              \
>     if (report_policy == ReportPolicy::DecodeError) {                                              \
>       callbacks_.onDecodeError();                                                                  \
>     }                                                                                              \
>     return absl::InvalidArgumentError(message);                                                    \
>   }
> ```
> 
> `utils.cc` also needs to access these marcos, but it cannot identify `callbacks_`, which defines in `decode.h`. Since `decode.h` already imported `utils.h`, if I try to make `utils.h` import `decode.h`, there will be circular dependency.

I was thinking 
```
#define RETURN_INVALID_ARG_ERR_IF_STATUS_NOT_OK(status, action)  \
  if (!status.ok()) {                      \
    action;                                \
    return status;                         \
  }
```
but seems to introduce more repetition. 

Let's not keep the dedicate macro for `onDecodeError` for now. Thanks for investigating on this

**Comment by Winbobob** [comment]
You mean "let's keep ..."?

**Comment by JuniorHsu** [comment]
> keep ..."?

Yes 🤦

**Comment by Winbobob** [comment]
The PR is ready for review again, cc @JuniorHsu , thanks!

**Review by JuniorHsu — APPROVED** [review]
I'm fine with this approach. Over to maintainers' round @mattklein123

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**